doc/cprover-manual/contracts-assigns.md

```diff
@@ -1,50 +1,144 @@
+of the function call or loop execution.
+
+Any memory that is locally stack- or heap-allocated during function or loop
+execution is by definition not part of the frame and always assignable by the
+function or the loop.
+
+## Syntax
 
 ```c
-__CPROVER_assigns(*identifier*, ...)
+__CPROVER_assigns(*targets*)
+```
+
+Where targets have the following syntax:
+
+```
+targets ::= conditional-target-group (';' conditional-target-group)* ';'?
+conditional-target-group ::= (condition ':')? target (',' target)*
+target ::= l-value
+        | __CPROVER_typed_target(l-value)
+        | __CPROVER_object_whole(pointer-typed-expression)
+        | __CPROVER_object_from(pointer-typed-expression)
+        | __CPROVER_object_upto(pointer-typed-expression,
+                                unsigned-integer-typed-expression)
 ```
 
-An _assigns_ clause allows the user to specify that a memory location may be written by a function. The set of locations writable by a function is the union of the locations specified by the assigns clauses, or the empty set of no _assigns_ clause is specified. While, in general, an _assigns_ clause could be interpreted with either _writes_ or _modifies_ semantics, this
-design is based on the former. This means that memory not captured by an
-_assigns_ clause must not be written within the given function, even if the
-value(s) therein are not modified.
+We use the _assigns_ interpretation for these locations (as opposed to _modifies_)
+meaning that locations not listed in the contract must not be assigned to
+by the function or the loop, even if they end up holding the same value as they
+held before the function call or before entering the loop.
 
-### Object slice expressions
+### Lvalue targets
 
-The following functions can be used in assigns clause to specify ranges of 
-assignable addresses.
+An l-value target `expr` with a complete type `expr_t` specifies that the range
+ of `sizeof(expr_t)` bytes starting at `&expr` is assignable.
 
-Given a pointer `ptr` pointing into some object `o`, `__CPROVER_object_from(ptr)` 
-specifies that all bytes starting from the given pointer and until the end of 
-the object are assignable:
+Similarly, given an l-value expression `expr` with a complete type `expr_t`,
+ `__CPROVER_typed_target(expr)` specifies that the range
+ of `sizeof(expr_t)` bytes starting at `&expr` is assignable:
 ```c
-__CPROVER_size_t __CPROVER_object_from(void *ptr); 
+__CPROVER_assignable_t __CPROVER_typed_target(expr_t expr);
 ```
 
-Given a pointer `ptr` pointing into some object `o`, `__CPROVER_object_from(ptr, size)` 
-specifies that `size` bytes starting from the given pointer and until the end of the object are assignable.
-The `size` value must such that `size <= __CPROVER_object_size(ptr) - __CPROVER_pointer_offset(ptr)` holds:
+To specify assignable memory locations
+that are interpreted as pointers by the program, one must use
+`__CPROVER_assigns(ptr)` or `__CPROVER_assigns(__CPROVER_typed_target(ptr))`.
```

Comment:
So how is this different from the above, at least for the case where `ptr` is an lvalue?